File: hw/morillo/u3/Hw29 Composite, Observer, Strategy & Template/TemplateMethodAplication/src/ec/edu/espe/composite/controller/CaffeineBeverage.java
package ec.edu.espe.composite.controller;

import javax.swing.JOptionPane;

/**
 *
 * @author David Morillo, JavaSquad, DCCO-ESPE
 */
public abstract class CaffeineBeverage {

    public void prepareRecipe() {
        boilWater();
        brew();
        pourInCup();

    }

    void boilWater() {
        JOptionPane.showMessageDialog(null, "Boiling Water");
    }

    abstract void brew();

    public void pourInCup() {
        JOptionPane.showMessageDialog(null, "Pouring into cup");

    }
    abstract void addCondiments();
    boolean wantsCondiments(){
        return wantsCondiments();
    }
}

File: hw/segarra/u3/HW29TemplateAndCompsite/HW29TemplateMethodJavaScript/ec.edu.espe.template/controller/CaffeineBeverage.js
class CaffeineBeverage {
    prepareRecipe() {
        this.boilWater();
        this.brew();
        this.pourInCup();
        if (this.wantsCondiments()) {
            this.addCondiments();
        }
    }

    boilWater() {
        console.log("Boiling water");
    }

    pourInCup() {
        console.log("Pouring into cup");
    }

    brew() {
        throw new Error("This method should be overridden!");
    }

    addCondiments() {
        throw new Error("This method should be overridden!");
    }

    wantsCondiments() {
        return true;
    }
}